repose-aggregator/components/filters/keystone-v2/src/main/scala/org/openrepose/filters/keystonev2/RequestHandler.scala

```diff
@@ -0,0 +1,524 @@
+                tokenTenants.find(reqTenant.equals) match {
+                  case Some(uriTenant) => Success(Some(uriTenant))
+                  case None => Failure(InvalidTenantException("Tenant from URI does not match any of the tenants associated with the provided token"))
+                }
+              case None => Failure(UnparseableTenantException("Could not parse tenant from the URI"))
+            }
+        }
+      } match {
+        case Some(Failure(e)) => Failure(e)
+        case Some(Success(uriTenant)) => Success(buildTenantVector(validToken.defaultTenantId, validToken.tenantIds, uriTenant))
+        case None => Success(buildTenantVector(validToken.defaultTenantId, validToken.tenantIds, None))
+      }
+    }
+  }
+
+  final def getEndpointsForToken(authenticatingToken: String, forToken: String): Try[EndpointsData] = {
+    val identityEndpoint = config.getIdentityService.getUri
+    /**
+     * Extract the user's endpoints from the endpoints call
+     * @param jsonString the Identity Endpoints call body
+     * @return a success or failure of a Vector[Endpoint] information
+     */
+    def extractEndpointInfo(jsonString: String): Try[EndpointsData] = {
+      import play.api.libs.functional.syntax._
+      import play.api.libs.json.Reads._
+      import play.api.libs.json._
+
+      implicit val endpointsReader = (
+        (JsPath \ "region").readNullable[String] and
+          (JsPath \ "name").readNullable[String] and
+          (JsPath \ "type").readNullable[String] and
+          (JsPath \ "publicURL").read[String]
+        )(Endpoint.apply _)
+
+      val json = Json.parse(jsonString)
+      //Have to convert it to a vector, because List isn't serializeable in 2.10
+      (json \ "endpoints").validate[Vector[Endpoint]] match {
+        case s: JsSuccess[Vector[Endpoint]] =>
+          val endpoints = s.get
+          val endpointsData = new EndpointsData(jsonString, endpoints)
+          Option(config.getCache) foreach { cacheSettings =>
+            // never null because configurationUpdated in KeystoneV2Filter
+            val timeout = offsetTtl(cacheSettings.getTimeouts.getEndpoints, cacheSettings.getTimeouts.getVariability)
+            datastore.put(s"$ENDPOINTS_KEY_PREFIX$forToken", endpointsData, timeout, TimeUnit.SECONDS)
+          }
+          Success(endpointsData)
+        case f: JsError =>
+          Failure(new IdentityCommuncationException("Identity didn't respond with proper Endpoints JSON"))
+      }
+    }
+
+    Try(akkaServiceClient.get(s"$ENDPOINTS_KEY_PREFIX$forToken",
+      s"$identityEndpoint${ENDPOINTS_ENDPOINT(forToken)}",
+      (Map(CommonHttpHeader.AUTH_TOKEN.toString -> authenticatingToken,
+        CommonHttpHeader.ACCEPT.toString -> MediaType.APPLICATION_JSON)
+        ++ traceId.map(CommonHttpHeader.TRACE_GUID.toString -> _)).asJava)) match {
+      case Success(serviceClientResponse) =>
+        serviceClientResponse.getStatus match {
```

Comment:
this set of checks is done a few times with basically identical responses other than on success maybe abstract and give it a function with the behavior wanted on success?